libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+
+  if (lastopt)
+    lastopt->next = optarg;
+  else
+    cmd->optargs = optarg;
+  cli_optarg_build_shortest(cmd->optargs);
+  retval = CLI_OK;
+
+CLEANUP:
+  if (retval != CLI_OK) {
+    cli_free_optarg(optarg);
+  }
+  return retval;
+}
+
+int cli_unregister_optarg(struct cli_command *cmd, const char *name) {
+  struct cli_optarg *ptr;
+  struct cli_optarg *lastptr;
+  int retval = CLI_ERROR;
+  // iterate looking for this option name, stoping at end or if name matches
+  for (lastptr = NULL, ptr = cmd->optargs; ptr && strcmp(ptr->name, name); lastptr = ptr, ptr = ptr->next) {
+    ;
+  }
+
+  // if ptr, then we found the optarg to delete
+  if (ptr) {
+    if (lastptr) {
+      // not first optarg
+      lastptr->next = ptr->next;
+      ptr->next = NULL;
+    } else {
+      // first optarg
+      cmd->optargs = ptr->next;
+      ptr->next = NULL;
+    }
+    cli_free_optarg(ptr);
+    cli_optarg_build_shortest(cmd->optargs);
+    retval = CLI_OK;
+  }
+  return retval;
+}
+
+void cli_unregister_all_optarg(struct cli_command *c) {
+  struct cli_optarg *o, *p;
+
+  for (o = c->optargs; o; o = p) {
+    p = o->next;
+    cli_free_optarg(o);
+  }
+}
+
+void cli_int_unset_optarg_value(struct cli_def *cli, const char *name) {
+  struct cli_optarg_pair **p, *c;
+  for (p = &cli->found_optargs, c = *p; *p;) {
+    c = *p;
+
+    if (!strcmp(c->name, name)) {
+      *p = c->next;
+      free_z(c->name);
+      free_z(c->value);
+      free_z(c);
+    } else {
+      p = &(*p)->next;
+    }
+  }
+}
+
+int cli_int_add_optarg_value(struct cli_def *cli, const char *name, const char *value, int allow_multiple) {
+  struct cli_optarg_pair *optarg_pair, **anchor;
+  int rc = CLI_ERROR;
+
+  for (optarg_pair = cli->found_optargs, anchor = &cli->found_optargs; optarg_pair;
+       anchor = &optarg_pair->next, optarg_pair = optarg_pair->next) {
+    // break if we found this name *and* allow_multiple is false
+    if (!strcmp(optarg_pair->name, name) && !allow_multiple) {
+      break;
+    }
+  }
+  // if we *didn't* find this, then allocate a new entry before proceeding
+  if (!optarg_pair) {
+    optarg_pair = (struct cli_optarg_pair *)calloc(1, sizeof(struct cli_optarg_pair));
+    *anchor = optarg_pair;
+  }
+  // set the value
+  if (optarg_pair) {
+    // name is null only if we didn't find it
+    if (!optarg_pair->name) optarg_pair->name = strdup(name);
+
+    // value may be overwritten, so free any old value.
+    if (optarg_pair->value) free_z(optarg_pair->value);
+    optarg_pair->value = strdup(value);
+
+    rc = CLI_OK;
+  }
+  return rc;
+}
+
+struct cli_optarg_pair *cli_get_all_found_optargs(struct cli_def *cli) {
+  if (cli) return cli->found_optargs;
+  return NULL;
+}
+
+char *cli_get_optarg_value(struct cli_def *cli, const char *name, char *find_after) {
+  char *value = NULL;
+  struct cli_optarg_pair *optarg_pair;
+
+  printf("cli_get_optarg_value entry - looking for <%s> after <%p>\n", name, (void *)find_after);
```

Comment:
Can you write debugging information to stderr instead of stdout?

Reply:
Woops.  Meant to pull that code - was there for testing.  Will rip that out.